README.rst
``autorepr``: makes civilized string representations
====================================================

.. image:: https://travis-ci.org/wolever/autorepr.svg?branch=master
    :target: https://travis-ci.org/wolever/autorepr

Now with Python 3 support!

Overview
--------

Python makes classes easy, but ``__repr__`` methods hard. Did you forget to
reference ``self`` again? Probably. Did have you thought to yourself "eh, this
class is real simple, it doesn't need a repr"? Without a doubt. Was production
taken down three times last week because your ``__str__`` returned unicode? ...
no? Maybe that's just me.

``autorepr`` makes it simple to build expressive, safe, and correct,
``__repr__``, ``__str__``, ``__unicode__``, and ``__bytes__`` methods in a
single line each.

With ``autorepr``, you get the repers you want, without worrying about the
fiddly bits (like encoding and decoding), leaving you to focus on your project:

.. code:: python

    >>> from autorepr import autorepr, autotext
    >>> class Person(object):
    ...     name = u"Alex ☃"
    ...     height = 123.456
    ...
    ...     __repr__ = autorepr(["name", "height:0.1f"])
    ...     __str__, __unicode__ = autotext("{self.name} ({self.height:0.0f} cm)")
    ...
    >>> p = Person()
    >>> repr(p)
    "<__main__.Person name=u'Alex \\u2603' height=123.5 at 0x...>"
    >>> unicode(p)
    u'Alex \u2603 (123 cm)'
    >>> str(p)
    'Alex \xe2\x98\x83 (123 cm)'


Installation
------------

::

    $ pip install autorepr


Usage
-----

``autorepr`` exposes two main functions:

- ``autorepr``, which builds a Python-esque ``__repr__`` string by passing
  either a ``str.format``-style string, or a list of attributes which should be
  included in a ``name=value`` list::

    autorepr(["name", "height:0.1f"]) -->
        "<pkg.Person name=u'Alex \u2603' height=123.5 at 0x...>"
    autorepr("{self.id} name={self.name!r}") -->
        "<pkg.Person 123 name=u'Alex \u2603' at 0x...>"

- ``autotext``, which uses ``autostr`` and ``autounicode`` to create
  ``__str__`` and ``__unicode__`` methods in a Python 2 + 3 friendly way::

    __str__, __unicode__ = autotext("{self.name} ({self.height!d} cm)") -->
        str: 'Alex \xe2\x98\x83 (123cm)'
        unicode: u'Alex \u2603 (123cm)'

And three secondary functions - ``autostr``, ``autounicode``, and
``autobytes`` - which build ``__str__``, ``__unicode__``, and ``__bytes__``
functions, respectively. The functions will do their best to avoid Unicode
encoding / decoding errors, and will generally Do The Right Thing, even if the
inputs aren't necessarily sensible.

Note: the examples shown here are Python 2, but everything works equally well
under Python 3.

.. code:: python

    >>> from autorepr import autorepr, autotext, autostr, autounicode
    >>> class Person(object):
    ...     name = u"Alex ☃"
    ...     height = 123.456
    ...
    ...     __repr__ = autorepr(["name", "height:0.1f"])
    ...     __str__, __unicode__ = autotext("{self.name} ({self.height:0.0f} cm)")
    ...
    >>> p = Person()
    >>> repr(p)
    "<__main__.Person name=u'Alex \\u2603' height=123.5 at 0x...>"
    >>> unicode(p)
    u'Alex \u2603 (123 cm)'
    >>> str(p)
    'Alex \xe2\x98\x83 (123 cm)'


Notice that ``autostr`` and ``autorepr`` (as called here through ``autotext``)
are intelligent about converting to/from unicode (decoding/encoding as UTF-8)
as necessary:

.. code:: python

    >>> p.name = u"unicode: ☃"
    >>> unicode(p)
    u'unicode: \u2603 (123 cm)'
    >>> str(p)
    'unicode: \xe2\x98\x83 (123 cm)'
    >>> p.name = 'utf-8 bytes: \xe2\x98\x83'
    >>> unicode(p)
    u'utf-8 bytes: \u2603 (123 cm)'
    >>> str(p)
    'utf-8 bytes: \xe2\x98\x83 (123 cm)'

*Note*: ``autostr`` and ``autorepr`` won't crash on invalid UTF-8 (for example,
if ``autounicode`` is asked to turn binary data into unicode), but the result
is *undefined* and may not be desirable.

Additional properties can be passed in as ``kwargs``, which will be called with
the instance as a parameter:

.. code:: python

    >>> name_with_len = autostr("{self.name} length={len}",
    ...                         len=lambda self: len(self.name))
    ...
    >>> p.name = 'Alex'
    >>> name_with_len(p)
    'Alex length=4'

This works with ``autorepr``'s list mode too:

.. code:: python

    >>> repr_with_len = autorepr(["name", "len"],
    ...                          len=lambda self: len(self.name))
    ...
    >>> repr_with_len(p)
    "<__main__.Person name='Alex' len=4 at 0x...>"

If a regular format string is passed to ``autorepr``, it will use that instead
of the generated string:

.. code:: python

    >>> repr_with_str = autorepr("{self.name!r}")
    >>> repr_with_str(p)
    "<__main__.Person 'Alex' at 0x...>"

And of course, if you don't want your ``__repr__`` to be wrapped in
``<ClassName ...>``, you can use ``autostr``:

.. code:: python

    >>> repr_with_autostr = autostr("Person({self.name!r})")
    >>> repr_with_autostr(p)
    "Person('Alex')"


Format specifications can also be passed to ``autorepr`` if the default of
``!r`` is undesirable (for example, truncating floats):

.. code:: python

    >>> with_fmt_spec = autorepr(["duration:0.1f", "addr:x", "type!s"],
    ...                          duration=lambda x: 123.456,
    ...                          addr=lambda x: 0xabc123,
    ...                          type=lambda x: "foo")
    >>> with_fmt_spec(None)
    '<....NoneType duration=123.5 addr=abc123 type=foo at 0x...>'


CHANGELOG.txt
0.3.0 (2017-03-14)
    * Python 3 support! (thanks for the prompting, @vlcinsky)
    * Added new `autotext` function to simplify creating __str__ and
      __unicode__ methods.

0.2.0 (2016-10-01)
    * Allow format specifications (ex: `:0.2f`) to be used in autorepr
    * Doctest README

0.1.1 (2016-01-27)
    * It sure would be nice if PyPI let me re-upload files

0.1.0 (2016-01-27)
    * Initial release.


test.py
# -*- coding: utf-8 -*-

import os
import doctest

from nose.tools import assert_equal
from nose_parameterized import parameterized

from autorepr import autostr, autounicode, autorepr

class Person(object):
    def __init__(self, name):
        self.name = name


ascii = Person("Alex")
uni = Person(u"☃")
bin = Person("\x00\xff")

@parameterized([
    (ascii.name, autostr, ascii),
    (unicode(ascii.name), autounicode, ascii),
    (uni.name.encode("utf-8"), autostr, uni),
    (uni.name, autounicode, uni),
    (bin.name, autostr, bin),
    (u"\x00\xff", autounicode, bin),
])
def test_encodings(expected, func, input):
    f = func("{self.name}")
    res = f(input)
    assert_equal(res, expected)
    assert_equal(type(res), type(expected))


@parameterized([
    ("'Alex'", autorepr("{self.name!r}")),
    ("name='Alex'", autorepr(["name"])),
    ("42", autorepr("{foo}", foo=lambda self: 42)),
    ("name='Alex' foo=42", autorepr(["name", "foo"], foo=lambda self: 42)),
])
def test_autorepr(expected, func):
    res = func(ascii)
    assert_equal("<%s.Person %s at 0x%x>" %(__name__, expected, id(ascii)), res)


def test_with_function_as_input():
    f = autounicode(autostr("{self.name} {foo}", foo=lambda x: 42))
    assert_equal(f(ascii), "Alex 42")

def test_readme_doctests():
    res = doctest.testfile("README.rst", optionflags=doctest.ELLIPSIS, encoding="utf-8")
    assert_equal(res.failed, 0)


if __name__ == '__main__':
    import nose
    nose.run(argv=[__file__])


tests.py
# -*- coding: utf-8 -*-
import doctest

from nose.tools import assert_equal
from nose.plugins.skip import SkipTest
from nose_parameterized import parameterized

from autorepr import autostr, autounicode, autorepr, text, to_text, to_bytes, PY2


if PY2:
    to_dunderstr = to_bytes
else:
    to_dunderstr = to_text

class Person(object):
    def __init__(self, name):
        self.name = name


ascii = Person("Alex")
uni = Person(u"☃")
bin = Person("\x00\xff")

@parameterized([
    (1, ascii.name, autostr, ascii),
    (2, text(ascii.name), autounicode, ascii),
    (3, to_dunderstr(uni.name), autostr, uni),
    (4, uni.name, autounicode, uni),
    (5, bin.name, autostr, bin),
    (6, u"\x00\xff", autounicode, bin),
])
def test_encodings(num, expected, func, input):
    f = func("{self.name}")
    res = f(input)
    assert_equal(res, expected)
    assert_equal(type(res), type(expected))


@parameterized([
    ("'Alex'", autorepr("{self.name!r}")),
    ("name='Alex'", autorepr(["name"])),
    ("42", autorepr("{foo}", foo=lambda self: 42)),
    ("name='Alex' foo=42", autorepr(["name", "foo"], foo=lambda self: 42)),
])
def test_autorepr(expected, func):
    res = func(ascii)
    assert_equal("<%s.Person %s at 0x%x>" %(__name__, expected, id(ascii)), res)


def test_with_function_as_input():
    f = autounicode(autostr("{self.name} {foo}", foo=lambda x: 42))
    assert_equal(f(ascii), "Alex 42")

def test_readme_doctests():
    if not PY2:
        raise SkipTest
    res = doctest.testfile("README.rst", optionflags=doctest.ELLIPSIS, encoding="utf-8")
    assert_equal(res.failed, 0)


if __name__ == '__main__':
    import nose
    nose.run(argv=[__file__])


tests/test_it.py
# -*- coding: utf-8 -*-
import doctest
import pytest
from autorepr import autostr, autounicode, autorepr

# Python 3 compatibility hack
try:
    unicode('')
except NameError:
    unicode = str


class Person(object):
    def __init__(self, name):
        self.name = name

ascii = Person("Alex")
uni = Person(u"☃")
bin = Person("\x00\xff")


@pytest.mark.parametrize("expected,func,input", [
    (ascii.name, autostr, ascii),
    (unicode(ascii.name), autounicode, ascii),
    (uni.name.encode("utf-8"), autostr, uni),
    (uni.name, autounicode, uni),
    (bin.name, autostr, bin),
    (u"\x00\xff", autounicode, bin),
])
def test_encodings(expected, func, input):
    f = func("{self.name}")
    res = f(input)
    assert res == expected
    assert type(res) == type(expected)


@pytest.mark.parametrize("expected,func", [
    ("'Alex'", autorepr("{self.name!r}")),
    ("name='Alex'", autorepr(["name"])),
    ("42", autorepr("{foo}", foo=lambda self: 42)),
    ("name='Alex' foo=42", autorepr(["name", "foo"], foo=lambda self: 42)),
])
def test_autorepr(expected, func):
    res = func(ascii)
    assert "<%s.Person %s at 0x%x>" % (__name__, expected, id(ascii)) == res


def test_with_function_as_input():
    f = autounicode(autostr("{self.name} {foo}", foo=lambda x: 42))
    assert f(ascii) == "Alex 42"


def test_readme_doctests():
    res = doctest.testfile("../README.rst",
                           optionflags=doctest.ELLIPSIS,
                           encoding="utf-8")
    assert res.failed == 0


autorepr.py
from __future__ import print_function

import sys
import types
import string

def show_py2(x):
    if isinstance(x, unicode):
        print(repr(x))
    elif isinstance(x, str):
        print('b' + repr(x))
    else:
        print(x)

def show_py3(x):
    if isinstance(x, str):
        print('u' + ascii(x))
    else:
        print(x)

if sys.version_info < (3, 0, 0):
    PY2 = True
    text = unicode
    bytes = str
    dundertext_name = "__unicode__"
    dunderbytes_name = "__str__"
    show = show_py2
else:
    PY2 = False
    text = str
    bytes = bytes
    dundertext_name = "__str__"
    dunderbytes_name = "__bytes__"
    show = show_py3

PY3 = not PY2


def to_bytes(obj, encoding='utf-8', **encode_args):
    r"""
    Returns a ``str`` of ``obj``, encoding using ``encoding`` if necessary. For
    example::

        >>> some_str = b"\xff"
        >>> some_unicode = u"\u1234"
        >>> some_exception = Exception(u'Error: ' + some_unicode)
        >>> show(to_bytes(some_str))
        b'\xff'
        >>> show(to_bytes(some_unicode))
        b'\xe1\x88\xb4'
        >>> show(to_bytes(some_exception))
        b'Error: \xe1\x88\xb4'
        >>> show(to_bytes([42]))
        b'[42]'

    See source code for detailed semantics.
    """
    # Note: On py3, ``b'x'.__str__()`` returns ``"b'x'"``, so we need to do the
    # explicit check first.
    if isinstance(obj, bytes):
        return obj

    # We coerce to unicode if '__unicode__' is available because there is no
    # way to specify encoding when calling ``str(obj)``, so, eg,
    # ``str(Exception(u'\u1234'))`` will explode.
    if isinstance(obj, text) or hasattr(obj, dundertext_name):
        # Note: unicode(u'foo') is O(1) (by experimentation)
        return text(obj).encode(encoding, **encode_args)

    return bytes(obj)

def to_text(obj, encoding='utf-8', fallback='latin1', **decode_args):
    r"""
    Returns a ``unicode`` of ``obj``, decoding using ``encoding`` if necessary.
    If decoding fails, the ``fallback`` encoding (default ``latin1``) is used.

    Examples::

        >>> show(to_text(b'\xe1\x88\xb4'))
        u'\u1234'
        >>> show(to_text(b'\xff'))
        u'\xff'
        >>> show(to_text(u'\u1234'))
        u'\u1234'
        >>> show(to_text(Exception(u'\u1234')))
        u'\u1234'
        >>> show(to_text([42]))
        u'[42]'

    See source code for detailed semantics.
    """

    # Note: on py3, the `bytes` type defines an unhelpful "__str__" function,
    # so we need to do this check (see comments in ``to_bytes``).
    if not isinstance(obj, bytes):
        if isinstance(obj, text) or hasattr(obj, dundertext_name):
            return text(obj)
        obj_str = bytes(obj)
    else:
        obj_str = obj

    try:
        return text(obj_str, encoding, **decode_args)
    except UnicodeDecodeError:
        return text(obj_str, fallback, **decode_args)


class SafeFormatter(string.Formatter):
    def __init__(self, coerce_func):
        self.coerce_func = coerce_func
        string.Formatter.__init__(self)

    def format_field(self, value, format_spec):
        res = string.Formatter.format_field(self, value, format_spec)
        return self.coerce_func(res)


class MagicBytesFormatter(string.Formatter):
    def format_field(self, value, format_spec):
        value = to_text(value)
        return string.Formatter.format_field(self, value, format_spec)


safe_bytes_formatter = SafeFormatter(to_bytes)
safe_text_formatter = SafeFormatter(to_text)
magic_bytes_formatter = MagicBytesFormatter()

def _autofmthelper(name, fmt, result_type, extra, postprocess=None):
    if isinstance(fmt, types.FunctionType):
        extra = fmt.extra
        fmt = fmt.fmt

    if result_type is text:
        safe_formatter = safe_text_formatter
        fmt = text(fmt)
    elif result_type is bytes:
        safe_formatter = safe_bytes_formatter
        fmt = bytes(fmt) if PY2 else text(fmt)
    elif result_type is repr:
        safe_formatter = None
        fmt = bytes(fmt) if PY2 else text(fmt)

    def fmtfunc(self):
        kwargs = dict((k, v(self)) for (k, v) in extra.items())
        kwargs["self"] = self
        if name == "__bytes__":
            result = magic_bytes_formatter.vformat(fmt, (), kwargs)
            result = to_bytes(result)
        else:
            try:
                result = fmt.format(**kwargs)
            except UnicodeError:
                if safe_formatter is None:
                    raise
                result = safe_formatter.vformat(fmt, (), kwargs)
        if postprocess is not None:
            result = postprocess(self, result)
        return result

    fmtfunc.__name__ = name
    fmtfunc.fmt = fmt
    fmtfunc.extra = extra
    return fmtfunc

def autounicode(fmt, **kwargs):
    """ Returns a simple ``__unicode__`` function::

        >>> class Person(object):
        ...     name = u"Alex \\u1234"
        ...     binary = b"\\x00\\x01\\x02"
        ...     __unicode__ = autounicode("{self.name} {self.binary}")
        ...
        >>> unicode(Person()) if PY2 else show(u'Alex \\u1234 \\x00\\x01\\x02')
        u'Alex \\u1234 \\x00\\x01\\x02'
        >>>
        """
    return _autofmthelper("__unicode__", fmt, text, kwargs)

def autostr(fmt, **kwargs):
    """ Returns a simple ``__str__`` function.
    
        On Python 2, an instance of ``str`` is returned with the same encoding
        behavior as ``autobytes`` (unicode is encoded as utf-8, bytes are left
        as bytes).

        On Python 3, an instance of ``text`` is returned with the same encoding
        behavior as ``autounicode`` (unicode is left as is, bytes are decoded
        as latin-1).

        For example::

        >>> class Person(object):
        ...     name = "Alex"
        ...     __str__ = autostr("{self.name}")
        ...
        >>> str(Person()) if PY2 else text(Person())
        'Alex'
        >>>
        """

    if isinstance(fmt, types.FunctionType):
        kwargs = fmt.extra
        fmt = fmt.fmt

    return _autofmthelper("__str__", fmt, bytes if PY2 else text, kwargs)

def autobytes(fmt, **kwargs):
    """ Returns a simple ``__bytes__`` function. Identical to ``autostr``
        on Python 2, and returns ``bytes`` on Python 3.
        
        Note: because Python 3's ``bytes`` type does not have a ``.format()``
        method we need to do a bit of magic here. Specifically, text will
        be utf-8 encoded and bytes will be left as-is.

        >>> class Person(object):
        ...     name = u"Alex \\u1234"
        ...     binary = b"\\x00\\x01\\x02"
        ...     __bytes__ = autobytes("{self.name} {self.binary}")
        ...
        >>> bytes(Person()) if PY3 else show('Alex \\xe1\\x88\\xb4 \\x00\\x01\\x02')
        b'Alex \\xe1\\x88\\xb4 \\x00\\x01\\x02'
        >>>
        """

    if isinstance(fmt, types.FunctionType):
        kwargs = fmt.extra
        fmt = fmt.fmt

    return _autofmthelper("__bytes__", fmt, bytes, kwargs)

def autorepr(fmt, **kwargs):
    """ Returns a simple ``__repr__`` function::

        >>> class Person(object):
        ...     name = "Alex"
        ...     __repr__ = autorepr(["name"])
        ...
        >>> repr(Person())
        "<autorepr.Person name='Alex' at 0x...>"
        >>> class Timestamp(object):
        ...     time = 123.456
        ...     __repr__ = autorepr(["time:0.01f"])
        ...
        >>> repr(Timestamp())
        '<autorepr.Timestamp time=123.5 at 0x...>'
        >>>
        >>> class Animal(object):
        ...     name = "Puppy"
        ...     __repr__ = autorepr(
        ...         "name={self.name!r} name_len={name_len!r}",
        ...          name_len=lambda self: len(self.name),
        ...     )
        ...
        >>> repr(Animal())
        "<autorepr.Animal name='Puppy' name_len=5 at 0x...>"
        >>>
        """
    if isinstance(fmt, list):
        fmt_bits = []
        for name in fmt:
            attr, _, fmt_spec = name.partition(":")
            attr, _, fmt_conv = attr.partition("!")
            fmt_suffix = (
                ":%s" %(fmt_spec, ) if fmt_spec else
                "!%s" %(fmt_conv or "r", )
            )
            fmt_bits.append("%s={%s%s}" %(
                attr,
                attr if attr in kwargs else "self.%s" %(attr, ),
                fmt_suffix,
            ))
        fmt = " ".join(fmt_bits)

    return _autofmthelper("__repr__", fmt, repr, kwargs, lambda self, result: (
        "<%s.%s %s at 0x%x>" %(
            self.__class__.__module__,
            self.__class__.__name__,
            result,
            id(self),
        )
    ))

def autotext(text_fmt, **kwargs):
    """ Returns a triple of ``(__str__, __unicode__)`` methods,
        for simple Python 2 / Python 3 compatibility::

        >>> class Person(object):
        ...     name = "Alex"
        ...
        ...     __str__, __unicode__ = autotext("{self.name}")
        ...
        >>> show(text(Person()))
        u'Alex'
    """

    return (
        autostr(text_fmt, **kwargs),
        autounicode(text_fmt, **kwargs),
    )

if __name__ == "__main__":
    import doctest
    __name__ = "autorepr"
    doctest.testmod(optionflags=doctest.ELLIPSIS)
